azimuth_schedule_operator/operator.py

```diff
@@ -120,3 +189,339 @@ async def schedule_check(body, namespace, **_):
+        if exc.response.status_code == 400:
+            try:
+                message = exc.response.json()["error_message"]
+            except (json.JSONDecodeError, TypeError, KeyError):
+                message = exc.response.text
+            raise BlazarLeaseCreateError(f"error creating blazar lease - {message}")
+        else:
+            raise
+
+
+def get_size_map(blazar_lease):
+    """Produce a map from requested size ID to reservation size ID."""
+    size_map = {}
+    for reservation in blazar_lease.get("reservations", []):
+        if (
+            reservation["resource_type"] == "flavor:instance"
+            and "resource_properties" in reservation
+        ):
+            properties = json.loads(reservation["resource_properties"])
+            size_map[properties["id"]] = reservation["id"]
+    return size_map
+
+
+async def get_size_name_map(cloud, size_map):
+    """Produce a size name map for the given size map."""
+    compute_client = cloud.api_client("compute")
+    flavor_names = {
+        flavor.id: flavor.name
+        async for flavor in compute_client.resource("flavors").list()
+    }
+    size_name_map = {}
+    for original_id, new_id in size_map.items():
+        try:
+            size_name_map[flavor_names[original_id]] = flavor_names[new_id]
+        except KeyError:
+            pass
+    return size_name_map
+
+
+async def update_lease_status_no_blazar(cloud, lease):
+    """Updates the lease status when Blazar is not used for the lease."""
+    if lease.spec.starts_at:
+        now = datetime.datetime.now(datetime.timezone.utc)
+        lease_started = now >= lease.spec.starts_at
+    else:
+        # No start date means start now
+        lease_started = True
+    if lease_started:
+        lease.status.set_phase(lease_crd.LeasePhase.ACTIVE)
+        lease.status.size_map = {
+            m.size_id: m.size_id for m in lease.spec.resources.machines
+        }
+        lease.status.size_name_map = await get_size_name_map(
+            cloud, lease.status.size_map
+        )
+    else:
+        lease.status.set_phase(lease_crd.LeasePhase.PENDING)
+
+
+@kopf.on.create(registry.API_GROUP, "lease")
+@kopf.on.resume(registry.API_GROUP, "lease")
+async def reconcile_lease(body, logger, **_):
+    lease = lease_crd.Lease.model_validate(body)
+
+    # Put the lease into a pending state as soon as possible
+    if lease.status.phase == lease_crd.LeasePhase.UNKNOWN:
+        lease.status.set_phase(lease_crd.LeasePhase.PENDING)
+        await save_instance_status(lease)
+
+    # Create a cloud instance from the referenced credential secret
+    secrets = await K8S_CLIENT.api("v1").resource("secrets")
+    cloud_creds = await secrets.fetch(
+        lease.spec.cloud_credentials_secret_name, namespace=lease.metadata.namespace
+    )
+    async with openstack.from_secret_data(cloud_creds.data) as cloud:
+        # If the lease has no end date, we don't attempt to use Blazar
+        if not lease.spec.ends_at:
```

Comment:
Nit: this could be outside the fetching of the creds, I think.

Reply:
Generating the size name map (for CAPI) requires talking to Nova to get the names for the flavor IDs that we have.

Reply:
(It was originally outside of the `async with` block before I realised that I needed that)